### PR TITLE
Keycloak 26.0.0

### LINKED ISSUE #1160: Keycloak 26
### Problem Statement

Keycloak 26 is currently not supported. Latest keycloak-config-cli version 6.1.5 prints warning ` Local keycloak-config-cli (6.1.5-25.0.1) and remote Keycloak (26.0.0) may not compatible.`

### Proposed Solution

_No response_

### Environment

- Keycloak Version: [e.g. 26.0.0]
- keycloak-config-cli Version: [e.g. 6.1.5]

### Additional information

_No response_

### Acceptance Criteria

_No response_

### PR DESCRIPTION
**Pull request by ma1uta**
**What this PR does / why we need it**:

Add support for Keycloak 26.0.0

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #1160

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

**Review by francis-pouatcha — CHANGES_REQUESTED** [review]
Can you please update the change log.

**Comment by thomasdarimont** [comment]
I just tested this branch and it worked for me.

However, I didn't test any KC26 specific configuration.

**Comment by thomasdarimont** [comment]
I also think we can drop the support for KC 18.x with the upgrade to KC26.

**Comment by SmithJosh** [comment]
Tested and this works for us as well. There was one bug we noticed though caused by this change in Keycloak 26 https://www.keycloak.org/docs/latest/release_notes/index.html#identity-providers-no-longer-available-from-the-realm-representation. The realm representation no longer contains identity provider mappers and so `keycloak-config-cli` tries to recreate them even if they already exist

**Comment by thomasdarimont** [comment]
I spent some time on this today but need to do other stuff now. I needed to adjust several other things as well (IdentityProviderMappers handling), see the latest commits in my branch.

Almost there:
image

**Comment by daviddavidgit** [comment]
I had a look at the branch of @thomasdarimont. Several tests are failing because Keycloak added constraints in Keycloak 26 when deleting an authentication flow: E.g. whena client overrides the default browser flow with a "custom flow" one can not just delete the "custom flow" later. So as soon as we import a config file with the "custom flow" defined, the strategy now is to first delete the "custom flow" and readd it. But this leads to error in Keycloak. I think we have to find a solution for that.

**Comment by ma1uta** [comment]
Thanks @thomasdarimont for help. Also it seems after https://github.com/keycloak/keycloak/commit/c4005d29f07d10eeda580d804a48fdfc9441f074 client-common-sync from https://github.com/keycloak/keycloak-client/ need to be update. They added the new field in RealmRepresentation but not synced the keycloak-client repo.

**Comment by ma1uta** [comment]
Also keycloak 26.0.4 [released](https://www.keycloak.org/2024/10/keycloak-2604-released.html).

**Comment by thomasdarimont** [comment]
@ma1uta keycloak-client 26.0.2 has been released that matches keycloak 26.0.4.

**Comment by ma1uta** [comment]
All pre-25 versions failed only with:
```
[ERROR] Failures: 
[ERROR]   ImportIdentityProvidersIT.shouldCreateOidcIdentityProvider:274 
Expected: is "example-client-secret"
     but: was "**********"
[ERROR]   ImportIdentityProvidersIT.shouldUpdateOidcIdentityProvider:323 
Expected: is "example-client-secret"
     but: was "**********"
[ERROR]   ImportIdentityProvidersIT.shouldUpdateOidcIdentityProviderWithDeleteAllMappers:628 
Expected: is "example-client-secret"
     but: was "**********"
[ERROR]   ImportIdentityProvidersIT.shouldUpdateOidcIdentityProviderWithMapper:372 
Expected: is "example-client-secret"
     but: was "**********"
[ERROR]   ImportIdentityProvidersIT.shouldUpdateOidcIdentityProviderWithReplacedMapper:564 
Expected: is "example-client-secret"
     but: was "**********"
[ERROR]   ImportIdentityProvidersIT.shouldUpdateOidcIdentityProviderWithUpdatedMapper:436 
Expected: is "example-client-secret"
     but: was "**********"
[ERROR]   ImportIdentityProvidersIT.shouldUpdateOidcIdentityProviderWithUpdatedMapperWithPseudoId:500 
Expected: is "example-client-secret"
     but: was "**********"
```

**Comment by jwklijnsma** [comment]
What is the eta for this do you need helping with testing / fix code ?

**Comment by leoandrea7** [comment]
Keycloak 26.0.5 was released

**Comment by jwklijnsma** [comment]
Identity Providers no longer available from the realm representation
As part of the improvements around the scalability of realms and organizations when they have many identity providers, the realm representation no longer holds the list of identity providers. However, they are still available from the realm representation when exporting a realm.

To obtain the query the identity providers in a realm, prefer using the /realms/{realm}/identity-provider/instances endpoint. This endpoint supports filters and pagination.

so next error 
Cannot create realm '************: HTTP 400 Bad Request{"errorMessage":"Identity provider mapper name must be unique per identity provider"}

File: pom.xml [hunk 1]
```diff
@@ -59,7 +59,9 @@
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
 
-        <keycloak.version>${keycloak.version}</keycloak.version>
+        <keycloak.version>${keycloak.version}}</keycloak.version>
```
Comment (bohmber) [inline pom.xml:62]: ${keycloak.version}} ? There are two `}`

Comment (ma1uta) [reply pom.xml]: this is a mistake

**Comment by jwklijnsma** [comment]
i am testing the fix cli but get this error Identity provider mapper name must be unique per identity provider it was working with older version?

**Comment by ma1uta** [comment]
Failed tests with `HTTP 500 Internal Server Error` because keycloak 26 throw exception `Cannot remove authentication flow, it is currently in use` and linked with https://github.com/keycloak/keycloak/pull/31875. Still investigating.

**Comment by bohmber** [comment]
Why keycloak should allow to delete and recreate a flow that is currently in use in a client?

**Comment by ma1uta** [comment]
There are two ways:

- don't allow delete flow which is used
- allow delete flow and replace with a standard one

But in previous versions of keycloak these tests succeed, need to find out what was changed in keycloak.

**Comment by bohmber** [comment]
`[org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-2) Uncaught server error: org.keycloak.models.ModelException: Cannot remove authentication flow, it is currently in use`
	`at org.keycloak.models.jpa.RealmAdapter.removeAuthenticationFlow(RealmAdapter.java:1460)`
	`at org.keycloak.models.cache.infinispan.RealmAdapter.removeAuthenticationFlow(RealmAdapter.java:1349)`
	`at org.keycloak.models.utils.KeycloakModelUtils.deepDeleteAuthenticationFlow(KeycloakModelUtils.java:967)`
	`at org.keycloak.services.resources.admin.AuthenticationManagementResource.deleteFlow(AuthenticationManagementResource.java:370)`
	`at org.keycloak.services.resources.admin.AuthenticationManagementResource$quarkusrestinvoker$deleteFlow_c7a3b2245d7e636a2fcab9cafd6ed82fb5d1f875.invoke(Unknown Source)`
	`at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)`
	`at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)`
	`at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)`
	`at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:635)`
	`at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)`
	`at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)`
	`at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)`
	`at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)`
	`at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)`
	`at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)`
	`at java.base/java.lang.Thread.run(Thread.java:1583)`

**Comment by bohmber** [comment]
The specific code is here in `AuthenticationFlowsImportService#recreateTopLevelFlow`
To update a flow the flow will be deleted and recreated. Before the deletion all references to this flow will be disabled/unasigned and afterwards enabled/asign again. This workaround is not done on clients. I think the check on client flows was not working < Keycloak 26. And I don't know why there is a difference in updating build-in flows and other flows. Build-in flows will be updated without this workaround.
```java
   /**
     * Deletes the top-level flow and all its executions and recreates them.
     */
    private void recreateTopLevelFlow(
            RealmImport realmImport,
            AuthenticationFlowRepresentation topLevelFlowToImport,
            AuthenticationFlowRepresentation existingAuthenticationFlow
    ) {
        AuthenticationFlowRepresentation patchedAuthenticationFlow = CloneUtil.patch(
                existingAuthenticationFlow, topLevelFlowToImport, "id"
        );

        if (existingAuthenticationFlow.isBuiltIn()) {
            throw new InvalidImportException(String.format(
                    "Unable to recreate flow '%s' in realm '%s': Deletion or creation of built-in flows is not possible",
                    patchedAuthenticationFlow.getAlias(), realmImport.getRealm()
            ));
        }

        UsedAuthenticationFlowWorkaroundFactory.UsedAuthenticationFlowWorkaround workaround = workaroundFactory.buildFor(realmImport);
        workaround.disableTopLevelFlowIfNeeded(topLevelFlowToImport.getAlias());

        authenticatorConfigImportService.deleteAuthenticationConfigs(realmImport, patchedAuthenticationFlow);
        authenticationFlowRepository.delete(realmImport.getRealm(), patchedAuthenticationFlow.getId());
        authenticationFlowRepository.createTopLevel(realmImport.getRealm(), patchedAuthenticationFlow);

        AuthenticationFlowRepresentation createdTopLevelFlow = authenticationFlowRepository.getByAlias(
                realmImport.getRealm(), topLevelFlowToImport.getAlias()
        );
        executionFlowsImportService.createExecutionsAndExecutionFlows(realmImport, topLevelFlowToImport, createdTopLevelFlow);

        workaround.resetFlowIfNeeded();
    }
```

**Comment by ma1uta** [comment]
@bohmber I found the original issue: https://github.com/keycloak/keycloak/issues/30707. So, I replace all overrides with temporary flow before deletion and restore after creation.

**Comment by bohmber** [comment]
@ma1uta The last failing test is something similar https://github.com/adorsys/keycloak-config-cli/pull/1162/commits/8f2f5dc5d3b65c06c660a3fb7b18265df605e8a2 in the file [44_update_realm_remove_authz_policy_realm-management.json](https://github.com/bohmber/keycloak-config-cli/blob/fcf3a3047fcf02252f6b1666e8f14a5378e7d0da/src/test/resources/import-files/clients/44_update_realm_remove_authz_policy_realm-management.json#L62)  the `realm-management` client has `authorizationServicesEnabled` with an empty resources section in authorizationSettings. The authorizationSettings will be deleted during update of the client. Later the code tries to update the deleted authorizationSettings. That is the `jakarta.ws.rs.NotFoundException: HTTP 404 Not Found`.

**Comment by ma1uta** [comment]
Keycloak create AuthorizationSettings in two cases:

- create resource
- `bearer only` and `public client` is false and `authorizationServicesEnabled` is true (https://github.com/keycloak/keycloak/blob/26.0.5/services/src/main/java/org/keycloak/services/resources/admin/ClientResource.java#L832) Also service account should be enabled (https://github.com/keycloak/keycloak/blob/26.0.5/services/src/main/java/org/keycloak/authorization/admin/ResourceServerService.java#L85)

**Comment by bohmber** [comment]
@ma1uta good point. That was exactly the code I was searching for the behavior

**Review by bohmber — APPROVED** [review]
Looks good for me. Tested it with Keycloak 26 and our configuration yesterday

**Review by francis-pouatcha — APPROVED** [review]

**Comment by jwklijnsma** [comment]
found still issue with edge image 

2024-11-21T13:17:54.804Z DEBUG 1 --- [           main] org.apache.http.wire                     : http-outgoing-0 >> "Host: keycloak-headless:8080[\r][\n]"
2024-11-21T13:17:54.804Z DEBUG 1 --- [           main] org.apache.http.wire                     : http-outgoing-0 >> "Connection: Keep-Alive[\r][\n]"
2024-11-21T13:17:54.804Z DEBUG 1 --- [           main] org.apache.http.wire                     : http-outgoing-0 >> "User-Agent: Apache-HttpClient/4.5.14 (Java/21.0.5)[\r][\n]"
2024-11-21T13:17:54.804Z DEBUG 1 --- [           main] org.apache.http.wire                     : http-outgoing-0 >> "[\r][\n]"
2024-11-21T13:17:54.811Z DEBUG 1 --- [           main] org.apache.http.wire                     : http-outgoing-0 << "HTTP/1.1 404 Not Found[\r][\n]"
2024-11-21T13:17:54.811Z DEBUG 1 --- [           main] org.apache.http.wire                     : http-outgoing-0 << "content-length: 30[\r][\n]"
2024-11-21T13:17:54.811Z DEBUG 1 --- [           main] org.apache.http.wire                     : http-outgoing-0 << "Content-Type: application/json[\r][\n]"
2024-11-21T13:17:54.811Z DEBUG 1 --- [           main] org.apache.http.wire                     : http-outgoing-0 << "Referrer-Policy: no-referrer[\r][\n]"
2024-11-21T13:17:54.811Z DEBUG 1 --- [           main] org.apache.http.wire                     : http-outgoing-0 << "Strict-Transport-Security: max-age=31536000; includeSubDomains[\r][\n]"
2024-11-21T13:17:54.811Z DEBUG 1 --- [           main] org.apache.http.wire                     : http-outgoing-0 << "X-Content-Type-Options: nosniff[\r][\n]"
2024-11-21T13:17:54.811Z DEBUG 1 --- [           main] org.apache.http.wire                     : http-outgoing-0 << "X-Frame-Options: SAMEORIGIN[\r][\n]"
2024-11-21T13:17:54.811Z DEBUG 1 --- [           main] org.apache.http.wire                     : http-outgoing-0 << "X-XSS-Protection: 1; mode=block[\r][\n]"
2024-11-21T13:17:54.811Z DEBUG 1 --- [           main] org.apache.http.wire                     : http-outgoing-0 << "[\r][\n]"
2024-11-21T13:17:54.811Z DEBUG 1 --- [           main] org.apache.http.wire                     : http-outgoing-0 << "{"error":"HTTP 404 Not Found"}"
2024-11-21T13:17:54.812Z DEBUG 1 --- [           main] org.apache.http.headers                  : http-outgoing-0 << HTTP/1.1 404 Not Found
2024-11-21T13:17:54.812Z DEBUG 1 --- [           main] org.apache.http.headers                  : http-outgoing-0 << content-length: 30
2024-11-21T13:17:54.812Z DEBUG 1 --- [           main] org.apache.http.headers                  : http-outgoing-0 << Content-Type: application/json
2024-11-21T13:17:54.812Z DEBUG 1 --- [           main] org.apache.http.headers                  : http-outgoing-0 << Referrer-Policy: no-referrer
2024-11-21T13:17:54.812Z DEBUG 1 --- [           main] org.apache.http.headers                  : http-outgoing-0 << Strict-Transport-Security: max-age=31536000; includeSubDomains
2024-11-21T13:17:54.812Z DEBUG 1 --- [           main] org.apache.http.headers                  : http-outgoing-0 << X-Content-Type-Options: nosniff
2024-11-21T13:17:54.812Z DEBUG 1 --- [           main] org.apache.http.headers                  : http-outgoing-0 << X-Frame-Options: SAMEORIGIN
2024-11-21T13:17:54.812Z DEBUG 1 --- [           main] org.apache.http.headers                  : http-outgoing-0 << X-XSS-Protection: 1; mode=block

**Pull request merged**